Just keep it here and run it first, for aligning all tables to the left side of the page

In [1]:
%%html
<style>
table {align:left;display:block} 
</style>

In [2]:
states = ["Sunny", "Cloudy", "Rainy"]

A = {
    "Sunny":  {"Sunny": 0.33, "Cloudy": 0.67, "Rainy": 0.00},
    "Cloudy": {"Sunny": 0.33, "Cloudy": 0.00, "Rainy": 0.67},
    "Rainy":  {"Sunny": 0.33, "Cloudy": 0.33, "Rainy": 0.33},
}

B = {
    "Sunny":  {"Walk": 1.0,  "Umbrella": 0.0},
    "Cloudy": {"Walk": 2/3,  "Umbrella": 1/3},
    "Rainy":  {"Walk": 1/3,  "Umbrella": 2/3},
}

pi = {"Sunny": 1/3, "Cloudy": 1/3, "Rainy": 1/3}

obs = ["Walk", "Umbrella", "Walk"]
T = len(obs)

V = [{s: 0.0 for s in states} for _ in range(T)]
bp = [{s: None for s in states} for _ in range(T)]

# Initialization
for s in states:
    V[0][s] = pi[s] * B[s][obs[0]]

# Recursion
for t in range(1, T):
    for curr in states:
        best_prev = None
        best_val = -1.0
        for prev in states:
            val = V[t-1][prev] * A[prev][curr]
            if val > best_val:
                best_val = val
                best_prev = prev
        V[t][curr] = best_val * B[curr][obs[t]]
        bp[t][curr] = best_prev

# Backtracking
last = max(states, key=lambda s: V[T-1][s])
path = [last]

for t in range(T-1, 0, -1):
    path.append(bp[t][path[-1]])

path.reverse()

print("Viterbi Table:")
for t in range(T):
    print(f"Day {t+1}: {V[t]}")

print("\nMost likely sequence:", path)
print("Best path probability:", V[T-1][path[-1]])


Viterbi Table:
Day 1: {'Sunny': 0.3333333333333333, 'Cloudy': 0.2222222222222222, 'Rainy': 0.1111111111111111}
Day 2: {'Sunny': 0.0, 'Cloudy': 0.07444444444444444, 'Rainy': 0.09925925925925924}
Day 3: {'Sunny': 0.03275555555555555, 'Cloudy': 0.021837037037037033, 'Rainy': 0.016625925925925926}

Most likely sequence: ['Cloudy', 'Rainy', 'Sunny']
Best path probability: 0.03275555555555555


# Problem 2 – Hidden Markov Model  
## Forward-Backtracking using the Viterbi Algorithm

---

## Problem Description

We are given:

- Initial state probabilities  
- Transition probabilities  
- Emission probabilities  

We observe the sequence:

\[
(Walk, \; Umbrella, \; Walk)
\]

Our task is to use the **Viterbi algorithm (Forward Recursion + Backtracking)** to determine the most likely hidden weather sequence.

---

## Model Setup

### Hidden States
\[
\{Sunny, \; Cloudy, \; Rainy\}
\]

### Initial Probabilities
We assume equal probability of starting in any state:

\[
\pi(Sunny) = \pi(Cloudy) = \pi(Rainy) = \frac{1}{3}
\]

---

### Transition Probabilities

| From → To | Sunny | Cloudy | Rainy |
|------------|--------|--------|--------|
| Sunny      | 0.33   | 0.67   | 0.00   |
| Cloudy     | 0.33   | 0.00   | 0.67   |
| Rainy      | 0.33   | 0.33   | 0.33   |

---

### Emission Probabilities

| Weather → Behavior | Walk | Umbrella |
|--------------------|--------|-----------|
| Sunny              | 1.0    | 0.0       |
| Cloudy             | 0.67   | 0.33      |
| Rainy              | 0.33   | 0.67      |

---

## Viterbi Algorithm

The Viterbi recurrence relation is:

\[
V_t(j) = P(o_t \mid j) \cdot \max_i \left( V_{t-1}(i) \cdot P(j \mid i) \right)
\]

Where:
- \( V_t(j) \) is the highest probability of any path ending in state \( j \) at time \( t \)
- \( P(o_t \mid j) \) is the emission probability
- \( P(j \mid i) \) is the transition probability

---

## Final Result

After performing forward recursion and backtracking:

\[
\boxed{Cloudy \rightarrow Rainy \rightarrow Sunny}
\]

Best path probability:

\[
0.032756
\]

This represents the most likely hidden weather sequence given the observations:
\[
(Walk, \; Umbrella, \; Walk)
\]
